### PR TITLE
add DisableTorchFunction that matches DisableTorchDispatch

### LINKED ISSUE #87990: Dispatch mode and torch function disable are inconsistent
### 🐛 Describe the bug

`no_dispatch` disables both mode and subclass (because it excludes the Python dispatch key, which controls for both)

`_C.DisableTorchFunction` only disables subclass (this behavior is explicitly tested for in `test_mode_notimplemented_loop`
and `test_disable_subclass_not_mode`)

Probably easiest fix is to rename DisableTorchFunction to DisableTorchFunctionSubclass and then add a mode disabler.

### Versions

master

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88219
* #88218

Closes #87990. This implements a new disable guard that matches DisableTorchDispatch (disables all subclasses and modes)

### CODE REVIEW

File: aten/src/ATen/PythonTorchFunctionTLS.h [hunk 1]
```diff
@@ -20,11 +22,15 @@ struct TORCH_API PythonTorchFunctionTLS {
 
  private:
   // The mode TLS is split into
-  //   - disabled_, which says whether or not to disable all torch function
-  //   modes
+  //   - disable_subclass_, which says whether or not to disable torch function
+  //     on subclasses
+  //   - disable_all_, which says whether or not to disable torch function on
+  //     both subclasses and modes. If this is true, it takes precedence over
+  //     disable_subclass
```
Comment (ezyang) [inline aten/src/ATen/PythonTorchFunctionTLS.h:29]: Turn this into an enum, OR rep disable subclass and mode separately (in both cases, make illegal states unrepresentable)

File: aten/src/ATen/PythonTorchFunctionTLS.cpp [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include <ATen/PythonTorchFunctionTLS.h>
 #include <c10/core/TensorImpl.h>
+#include <iostream>
```
Comment (ezyang) [inline aten/src/ATen/PythonTorchFunctionTLS.cpp:3]: debuggin

File: aten/src/ATen/PythonTorchFunctionTLS.cpp [hunk 3]
```diff
@@ -43,7 +43,8 @@ const PythonTorchFunctionTLS& PythonTorchFunctionTLS::get_state() {
 }
 
 bool torch_function_mode_enabled() {
-  return PythonTorchFunctionTLS::stack_len() > 0;
+  return PythonTorchFunctionTLS::get_disabled_state() != TorchFunctionDisabledState::ALL_DISABLED &&
```
Comment (ezyang) [inline aten/src/ATen/PythonTorchFunctionTLS.cpp:46]: it seems better to factor this into a helper function that tests if modes are enabled... what if you add another entry to the enum?

File: torch/csrc/utils/disable_torch_function.cpp [hunk 4]
```diff
@@ -34,20 +35,23 @@ void set_disabled_torch_dispatch_impl(PyObject* value) {
...
     PyObject* self,
     PyObject* unused) {
-  ((DisableTorchFunctionSubclass*)self)->old_state =
-      at::impl::PythonTorchFunctionTLS::is_disabled();
-  at::impl::PythonTorchFunctionTLS::set_disabled(true);
+  const auto old_state = at::impl::PythonTorchFunctionTLS::get_disabled_state();
+  ((DisableTorchFunctionSubclass*)self)->old_state = old_state;
+  if (old_state == at::impl::TorchFunctionDisabledState::ENABLED) {
+    at::impl::PythonTorchFunctionTLS::set_disabled_state(
+        at::impl::TorchFunctionDisabledState::SUBCLASSES_DISABLED);
```
Comment (ezyang) [inline torch/csrc/utils/disable_torch_function.cpp:48]: Ditto with helper

**Review by ezyang — APPROVED** [review]

**Comment by izaitsevfb** [comment]
Reverting this PR as it's base PR causes internal failures due to the BC-breaking chage.

See D41211901 and #88218

**Pull request closed without merging**